### PR TITLE
add deployment status tracking for linux web apps

### LINKED ISSUE #3260: [Feature Request] - Integrate Deployment Status API in 'azd up' and 'azd deploy' for Azure App Service
Azure App Service has a new "deployment status" API that provides a lot more insight and stateful information for all the things that happen server side after binary copy is complete and before the app is ready to serve the first request.

This is the documentation for the API
https://learn.microsoft.com/en-us/rest/api/appservice/web-apps/get-production-site-deployment-status?view=rest-appservice-2022-03-01&tabs=HTTP

Integrating this with azd will give more insights into the web app deployment - both for Build and Runtime states. We have already integrated the API with Azure CLI and this is the blog for it.
https://azure.github.io/AppService/2024/01/10/Deployment-Status-API-CLI.html

Please also refer to https://github.com/Azure/azure-dev/issues/2669.

### PR DESCRIPTION
**Pull request by hemarina**
fix #3260, build on top of https://github.com/Azure/azure-dev/pull/3763

This PR fixes:
- [x] Add deployment status tracking for Linux web apps using[ kudo api](https://learn.microsoft.com/en-us/rest/api/appservice/web-apps/get-production-site-deployment-status?view=rest-appservice-2022-03-01&tabs=Go#deploymentbuildstatus)
- [x] Fix flaky 404 Resouces Not Found error that is caused by temp deployment id in kudo api
- [x] Silence build failure caused by retries inside go sdk will be fixed in separate PR https://github.com/Azure/azure-dev/pull/3985
- [x] Add detail status message for `armappservice.DeploymentBuildStatus`

### CODE REVIEW

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:155]: Here we're constructing the `arm.ClientOptions` from scratch. I think we should instead be inheriting the arm.ClientOptions default settings that is passed in during `NewZipDeployClient` instead of constructing one completely, I'll push a change shortly that introduces `c.armClientOptions`, but I'll leave it up to you here to wire it up.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:199]: It may look a little cleaner if we use the retry options for `azure-sdk-for-go`. I believe the corresponding change would look like the following for your use case here. Note that we use `policy.WithRetryOptions` to create a `context.Context` that only affects the current request, and not the overall client.

```golang
	retryCtx := policy.WithRetryOptions(ctx, policy.RetryOptions{
		MaxRetries: 4,
		StatusCodes: append([]int{
			http.StatusRequestTimeout,      // 408
			http.StatusTooManyRequests,     // 429
			http.StatusInternalServerError, // 500
			http.StatusBadGateway,          // 502
			http.StatusServiceUnavailable,  // 503
			http.StatusGatewayTimeout,      // 504
		}, http.StatusNotFound), // 404
	})
	// nolint:lll
	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
	poller, err := client.BeginGetProductionSiteDeploymentStatus(retryCtx, resourceGroup, appName, deploymentStatusId, nil)
	if err != nil {
		return nil, fmt.Errorf("getting deployment status: %w", err)
	}
```

**Review by weikanglim — APPROVED** [review]
LGTM! I just left two minor comments to slightly simplify the implementation. Thanks for seeing this one through!

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [reply cli/azd/pkg/azsdk/zip_deploy_client.go:155]: I pushed the change for `c.armClientOptions` that is ready for uptake.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (vhvb1989) [reply cli/azd/pkg/azsdk/zip_deploy_client.go:196]: the call to `client.BeginGetProductionSiteDeploymentStatus` does not manual retries. It can use the default sdk client retries.  This API is not the one sending the zip file to Azure.

Comment (vhvb1989) [reply cli/azd/pkg/azsdk/zip_deploy_client.go:196]: We need manual retry for the call in line 101:
https://github.com/Azure/azure-dev/pull/3965/files/9ad7904c603b8e87a447789a02920b09651b6e2b#diff-33f0d8e97cd66b4e5eafa3c7ad76e09444aeedf30441eb8d993c18e4588be8e5R101

Comment (weikanglim) [reply cli/azd/pkg/azsdk/zip_deploy_client.go]: FYI, while recently going through the implementation for `api/publish`, I did discover that `azure-sdk-for-go` handles retries for streaming body content perfectly fine. We just need to pass an `io.ReadSeekCloser` and use `request.SetBody`. You can see it working [here](https://github.com/Azure/azure-dev/pull/3923/files#diff-33f0d8e97cd66b4e5eafa3c7ad76e09444aeedf30441eb8d993c18e4588be8e5). After this lands, I'll fix the zip deploy logic here.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:76]: Curious about the logic here to *only* increase to 4 if the retry attempt is 3. I wonder if it's better to always set it to some retry attempt we care about. It's okay even if the options passed in had a higher number of attempts, we just want to ensure we know how long our request executes.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:166]: I think this [comment](https://github.com/Azure/azure-dev/pull/3965#discussion_r1625033368) may not be fully addressed yet, we're still using the `retry` library here:

> It may look a little cleaner if we use the retry options for `azure-sdk-for-go`. I believe the corresponding change would look like the following for your use case here. Note that we use `policy.WithRetryOptions` to create a `context.Context` that only affects the current request, and not the overall client.
> 
> ```go
> 	retryCtx := policy.WithRetryOptions(ctx, policy.RetryOptions{
> 		MaxRetries: 4,
> 		StatusCodes: append([]int{
> 			http.StatusRequestTimeout,      // 408
> 			http.StatusTooManyRequests,     // 429
> 			http.StatusInternalServerError, // 500
> 			http.StatusBadGateway,          // 502
> 			http.StatusServiceUnavailable,  // 503
> 			http.StatusGatewayTimeout,      // 504
> 		}, http.StatusNotFound), // 404
> 	})
> 	// nolint:lll
> 	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
> 	poller, err := client.BeginGetProductionSiteDeploymentStatus(retryCtx, resourceGroup, appName, deploymentStatusId, nil)
> 	if err != nil {
> 		return nil, fmt.Errorf("getting deployment status: %w", err)
> 	}
> ```

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:171]: "// 404 - Retry if the resource is not found as deployment id doesn't match with temp deployment id"

Without any context, I'm not getting the full details of why we would retry for 404s out of this comment. I'm not sure if this matches your expectations, but I think the reason would be `// We get temporary 404s when the ARM API hasn't fully committed the deployment record with the given deployment ID`.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [inline cli/azd/pkg/azsdk/zip_deploy_client.go:278]: I think this may be a problem, as the logic ends up swallowing the `err` returned here, and would return `nil` as a success?

**Review by weikanglim — COMMENTED** [review]
Did another pass based on the latest changes. Overall still looks good, but there are some comments we probably do want to address.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [reply cli/azd/pkg/azsdk/zip_deploy_client.go:168]: The main reason why I would choose azure-sdk-for-go's built-in retry is to avoid complexity (what I meant by "it may look a little cleaner"). We have a retry mechanism built in into the SDK, and we should opt for using it instead of something else if possible. 

When we use different retry mechanisms, we need to guarantee that we do not create unexpected retry loops due to retries on top of retries, which has been a source of bugs in many software. Stacking retries easily changes the behavior the retry delay to exponential.

File: cli/azd/pkg/azsdk/zip_deploy_client.go
Comment (weikanglim) [reply cli/azd/pkg/azsdk/zip_deploy_client.go:77]: I think the code here should inherit default settings from the application, and apply specific settings it cares about. Retries is one of those settings it cares about, and it should have complete control over the desired value. 

I also now notice that we reverted the [previous change](https://github.com/Azure/azure-dev/pull/3965/commits/bb7c9327d9e3703696046e32e78a581ea1f3cd13), and this is now modifying the passed in `*arm.ClientOptions` which is unexpected.

**Pull request merged**
